File: packages/flow-xml/src/actions/Untag.java
/*
 * Copyright © 2013-2025 Metreeca srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.metreeca.flow.xml.actions;

import com.metreeca.flow.http.FormatException;

import org.w3c.dom.*;

import java.io.ByteArrayInputStream;
import java.util.Locale;
import java.util.Optional;
import java.util.function.Function;

import static com.metreeca.flow.xml.XPath.normalize;
import static com.metreeca.flow.xml.formats.HTML.html;

import static java.nio.charset.StandardCharsets.UTF_8;

/**
 * X/HTML to Markdown conversion.
 *
 * <p>Converts an X/HTMl document to a markdown-based plain text representation.</p>
 */
public final class Untag implements Function<Node, String> {

    /**
     * Converts an X/HTMl document to a markdown-based plain text representation.
     *
     * @param document the content of the X/HTML document
     *
     * @return the markdown-based plain text representation of {@code document} or the original {@code document}
     * contents if unable to parse it as an X/HTML document
     *
     * @throws NullPointerException id {@code document } is {@code null}
     */
    public static String untag(final String document) {

        if ( document == null ) {
            throw new NullPointerException("null document");
        }

        try {

            return new Untag().apply(html(new ByteArrayInputStream(document.getBytes(UTF_8)), UTF_8, ""));

        } catch ( final FormatException e ) {

            return document;

        }
    }


    @Override public String apply(final Node element) {
        return element == null ? "" : new Builder().format(element).toString();
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private static final class Builder {

        private int indent;
        private boolean space;

        private final StringBuilder builder=new StringBuilder(100);


        private Builder format(final NodeList nodes) {

            for (int i=0, n=nodes.getLength(); i < n; ++i) {
                format(nodes.item(i));
            }

            return this;
        }

        private Builder format(final Node node) {
            return node instanceof Document ? format((Document)node)
                    : node instanceof Element ? format((Element)node)
                    : node instanceof Text ? format((Text)node)
                    : this;
        }

        private Builder format(final Document document) {

            document.normalize();

            format(document.getDocumentElement());

            return this;
        }

        private Builder format(final Element element) {
            switch ( element.getTagName().toLowerCase(Locale.ROOT) ) {

                case "h1":

                    return append("# ").append(normalize(element.getTextContent())).feed();

                case "h2":

                    return append("## ").append(normalize(element.getTextContent())).feed();

                case "h3":

                    return append("### ").append(normalize(element.getTextContent())).feed();

                case "p":
                case "div":
                case "section":

                    return format(element.getChildNodes()).feed();

                case "ul":
                case "ol":

                    return indent().format(element.getChildNodes()).outdent().feed();

                case "li":

                    return append("- ").format(element.getChildNodes()).wrap();

                case "br":

                    return append("  ").wrap();

                case "hr":

                    return append("---").feed();

                case "a":

                    return append("[").format(element.getChildNodes()).append("]").append("(")
                            .append(element.getAttribute("href")).append(")");

                case "img":

                    return append("![").format(element.getAttributeNode("alt")).append("]").append("(")
                            .append(element.getAttribute("src")).append(")");

                case "strong":
                case "b":

                    return append("**").append(normalize(element.getTextContent())).append("**");

                case "em":
                case "i":

                    return append("*").append(normalize(element.getTextContent())).append("*");

                case "script": // preserve JSON-LD metadata

                    return "application/ld+json".equals(element.getAttribute("type"))
                            ? append("```json\n").append(normalize(element.getTextContent())).append("\n```").feed()
                            : this;

                case "head":
                case "style":

                    return this;

                default:

                    return format(element.getChildNodes());

            }
        }

        private Builder format(final Text text) {

            Optional.ofNullable(text.getNodeValue())
                    .map(s -> normalize(s, true)) // preserve border whitespace to handle misplaced emphasis tags
                    .ifPresent(this::append);

            return this;
        }


        private Builder append(final String string) {

            for (int n=0; n < string.length(); ++n) {

                final char c=string.charAt(n);

                switch ( c ) {

                    case ' ', '\t' -> space=true;

                    case '\n' -> {

                        builder.append(c);

                        space=false;
                    }

                    case '\r' -> space=false;

                    default -> {

                        if ( space ) {
                            builder.append(' ');
                        }

                        builder.append(c);

                        space=false;
                    }
                }
            }

            return this;
        }


        private Builder indent() {

            if ( indent++ == 0 ) { feed(); }

            return this;
        }

        private Builder outdent() {

            indent--;

            return this;
        }

        private Builder feed() {

            if ( builder.length() > 1 && builder.charAt(builder.length()-1) != '\n' ) {
                builder.append('\n');
            }

            if ( builder.length() > 2 && builder.charAt(builder.length()-2) != '\n' ) {
                builder.append('\n');
            }

            space=false;

            return this;
        }

        private Builder wrap() {

            if ( builder.length() > 1 && builder.charAt(builder.length()-1) != '\n' ) {
                builder.append('\n');
            }

            space=false;

            return this;
        }


        @Override public String toString() {
            return builder.toString();
        }

    }

}

File: packages/gear-xml/src/untag.ts
/*
 * Copyright © 2026 Metreeca srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Markdown } from "@metreeca/core/strings";
import type { Task } from "@metreeca/flow";
import { map } from "@metreeca/flow/tasks";
import type { AnyNode } from "domhandler";
import { process } from "./untag.core.js";


/**
 * Creates a markdown renderer.
 *
 * The generated task converts a feed of parsed X/HTML trees into a feed of markdown text, one rendering per tree, so
 * that a consumer reading prose, a language model among them, works on text rather than on markup.
 *
 * A tree holding neither content nor a title is converted to an empty string, so that renderings stay aligned with the
 * trees they were drawn from.
 *
 * Where a tree states a title, its rendering opens with a YAML frontmatter block stating it, so that a consumer reads
 * the page the text belongs to alongside the text itself. The title is the first `title` element the tree states
 * outside the framing a reader is not after, so that the caption of an embedded object is not mistaken for it, and it
 * is written as a quoted scalar, so that the punctuation a headline carries doesn't unsettle the block. Where a tree
 * states no title, or one carrying no text, its rendering opens with the content.
 *
 * Elements are rendered as follows, names matched as the tree carries them, case insensitively:
 *
 * - `h1`, `h2`, `h3` — a heading of the matching level, set off by a blank line, left out where it carries no text
 * - `p`, `section`, `article` — the content, set off by a blank line
 * - `div` — the content, set off by a blank line where the element states text of its own or wraps a lone element,
 *   whitespace aside, and closed by a line break otherwise, so that a field reads as a paragraph while the wrappers a
 *   page is laid out with don't split its content into blocks of their own
 * - `ul`, `ol` — a list set off from the surrounding content by a blank line, ordered lists marked as unordered ones
 * - `li` — an item marked with `-`, indented by two spaces for each enclosing list beyond the outermost, its content
 *   opening on the line the marker is written on, however the item lays it out, left out where it carries neither text
 *   nor an image
 * - `br` — a line break, two of them laying down the blank line a paragraph is often split with, a longer run
 *   saturating at that blank line and a run opening the text dropped
 * - `hr` — a thematic break, set off by a blank line
 * - `a` — a link to the `href` stated, labelled by the content, left out where it carries neither text nor an image
 * - `img` — an image reference to the `src` stated, labelled by the `alt` text
 * - `strong`, `b` — strong emphasis, the whitespace bordering the content written outside the markers, as markers
 *   padded with it read as text rather than as emphasis, left out where it carries no text, though the space it holds
 *   is kept
 * - `em`, `i` — emphasis, laid out as strong emphasis is
 * - `script` — a fenced `json` block, if the type is `application/ld+json`, set off by a blank line; nothing
 *   otherwise
 * - `head`, `style`, `title` — nothing, the title being stated by the frontmatter instead
 *
 * Every other element contributes its content, the `html` and `body` a page is wrapped in among them, so that the
 * wrappers a page is built from leave no trace of their own. A link or an item is kept for the content a reader is
 * shown, the caption a graphic states inside its own markup counting for nothing, so that a decorative link leaves no
 * empty label behind.
 *
 * Character data is rendered with runs of spaces, control characters and typographic separators, the no-break space
 * among them, collapsed to a single space, whatever the markup lays out; a run bordering a text node is kept, so that
 * emphasis misplaced with respect to the surrounding spaces doesn't run words together. A comment carries no text but
 * counts as a space, so that the markers a framework leaves between elements keep the words on either side apart, as
 * do the fields a page lays out side by side. Text bordering an element runs into it as stated, so that a word split
 * across an element and the text beside it is not broken apart. A space never opens a line or closes a link label, so
 * that the whitespace a page is laid out with doesn't reach the text; leading and trailing whitespace is stripped from
 * each rendering.
 *
 * > [!NOTE]
 * >
 * > - **Incremental**: each rendering is emitted as soon as its tree is drawn, so the feed produced runs dry as the
 * >   feed drawn from does and an endless source is read as long as it is consumed.
 * > - **Materialising**: a rendering is assembled in memory as its tree is walked, so peak memory use is about the
 * >   size of the largest tree rather than of the feed.
 * > - **Stateless**: every tree is converted on its own, so the outcome is unaffected by how the feed is split across
 * >   nested feeds or runs.
 *
 * @returns A task converting a feed of parsed X/HTML trees into a feed of markdown text
 *
 * @throws {Error} While the feed is consumed, whatever the source reports while producing trees
 *
 * @see {@link https://spec.commonmark.org/ CommonMark Spec}
 * @see {@link https://json-ld.org/ JSON-LD}
 *
 * @group Factories
 */
export function untag(): Task<AnyNode, Markdown> {

	return map(process);

}
